Rework all QGIS boundcrs related logic

nyalldawson opened the pull request:
Trying to handle this on the QGIS side is dangerous -- we need to instead just trust what we get from proj and don't try to second-guess the results!

Potentially, this may cause some regressions, as boundcrs objects which previously were identified as known CRSes (e.g. EPSG:...) will no longer be, and will instead become user crses. It's both a cosmetic issue and potentially a source of projection errors (e.g. when a GDA2020 boundcrs isn't converted to the source crs then the available coordinate operations miss the grid based methods). But the alternative is that we incorrectly equate boundcrses with their underlying source crses, which also leads to incorrect projection result, and a lot of projection related logic in QGIS.

So this PR goes with the first approach, and drops all the QGIS side logic relating to these (on proj 6 builds only!!). Wherever we encounter CRSes which are incorrectly treated as user crses instead of standard auth:id CRSes, we can file tickets at proj and get the issues fixed upstream...

nyalldawson committed:
Don't use a boundCRS definition as the WKT definition of EPSG:4326
on proj 6 builds

See https://github.com/OSGeo/PROJ/issues/1804

nyalldawson committed:
Don't equate a BoundCRS with an identified SourceCRS as the SourceCRS when identifying
CRSes

May result in some CRSes which were previously identified as known standard CRSes
now being shown as User CRSes, but they aren't the same to we can't equate them.

If previously identified CRSes are showing as User CRSes now, then we should
investigate these upstream in the proj or gdal libraries, and get all fixes
in place upstream rather then trying to do this in QGIS code (which is
inevitably going to end up with an unmaintainable mess)

nyalldawson committed:
Don't incorrectly flag unknown CRS created from proj strings as invalid
on proj 6+ builds

nyalldawson committed:
Fix incorrectly ellipsoid acronym returned when creating a CRS from a custom non-standard proj string on proj 6+ builds

nyalldawson committed:
Remove a downstream hack

nyalldawson committed:
Fix projection mismatch when reading very old projects on proj 6 builds

We should always prioritise an exact auth:id means of constructing
CRSes whenever it's available, instead of just starting with a proj string.

nyalldawson committed:
Don't automatically create boundcrs proj string representations of CRSes

Just let proj decide when and if these are needed

nyalldawson committed:
Remove more proj4 related hacks from proj6 builds

nyalldawson committed:
Ensure that proj strings in different parameter order are matched
to user CRS on proj 6 builds

nyalldawson committed:
Fix matching of CRS by WKT to user CRSes defined by WKT, including
parameter-order independant matches

nirvn commented on src/core/qgscoordinatereferencesystem.cpp:
```diff
@@ -824,12 +824,19 @@ bool QgsCoordinateReferenceSystem::createFromProj4( const QString &proj4String )
-  // (and if any of this has value, take it up with the proj project. That's where it belongs)
-  // I ***think*** this is safe to disable for proj 6.1.1 and above
-#if 1
+#if PROJ_VERSION_MAJOR>=6
```
Are you sure you meant for it to be PROJ_VERSION_MAJOR>=6 and not <6 ?

nyalldawson replied:
No the logic below that is new proj 6 specific code, the rest was moved to an #else

nirvn replied:
Ah got it, sorry for the noise.

nyalldawson committed:
Spelling

nyalldawson committed:
Rip off a band aid, and deprecate all "proj4" calls from the API

Alias these across to new "proj" methods, which don't include the
version number.

nyalldawson committed:
Deprecate methods for constructing CRS from Postgis srids

Constructing CRS using Postgis srids is highly discouraged,
and instead CRSes should always be constructed using auth:id
codes or WKT strings.

QGIS 4.0: The logic should be isolated into the postgres
provider alone, and not exposed to stable API

nyalldawson committed:
Deprecate most methods which construct CRSes from internal CRS ids

These should NOT be used, and auth:code or WKT definitions used instead.

Unfortunately some use of these methods are very heavily entangled around
other code, so we can't deprecate all of them until 4.0

nyalldawson committed:
Never auto-save unknown CRSes encountered in the wild

Previously whenever an unknown CRS (i.e. something which couldn't
be matched a standard or user crs) was encountered then it would
automatically be saved to the user's custom CRS database.

This meant that over time, the custom CRS database would fill
with dozens of random entries. Even a clean QGIS install would
pollute the custom projection list with a bunch of automatically
created random CRS definitions!

Instead, never auto save CRSes to the custom projection database.
Only projections which users have manually created via the custom
projection dialog should ever be available as User CRSes.

nyalldawson committed:
Cleanup handling of recent CRS list, add tests

nyalldawson committed:
Add a method for retrieving a user friendly descriptive string for
a CRS

And use this wherever we show CRS values to users

nyalldawson committed:
More user of userFriendlyIdentifier

nyalldawson committed:
Update tests

nyalldawson committed:
Fix build, tests on proj < 6

nyalldawson committed:
Replace deprecated members

nyalldawson committed:
Update test results to reflect newer GDAL behavior

nyalldawson committed:
Revert "Don't automatically create boundcrs proj string representations of CRSes"

Following discussion with GDAL maintainers, it's been agreed that the consistent logic
to use across projects should be:

1. Proj strings: export as bound CRS, and treat a boundcrs match to a known CRS as equivalent
to the underlying source CRS

2. WKT strings: export without change, not as bound CRS, and DON'T treat boundcrs matches
as equivalent to the underlying source CRS

nyalldawson committed:
Update test results

nyalldawson committed:
Partial revert BoundCRS identification change

Sync behavior with the new GDAL behavior, where a CRS defined using
a proj string alone will be considered equivalent to the underlying
source crs.

nyalldawson committed:
New projection dialog proj 6 tweaks and revisions

nyalldawson committed:
Fix build on proj <6

nyalldawson commented:
This has become a mega PR, which should be the bulk of the remaining changes required to fix all outstanding proj 6 related issues.

I've also:
- aliases the "proj4" members from the QgsCoordinateReferenceSystem api and deprecated the original ones (since proj4 isn't correct, it's "proj")
- deprecated most of the unsafe methods of constructing CRS objects, i.e. those which rely on internal crs ids or postgis srids. I haven't been able to completely deprecate and remove all the use of all the methods as some are still very deeply ingrained in other parts of the code. I'll keep chipping away at this over time though
- synchronized the logic to provide a unified logic with gdal, based on https://github.com/OSGeo/gdal/pull/2113
- removed all forced saving of unknown crs as user crs, and left it so that only crs which users have actually manually created themselves are saved as user crs

nyalldawson committed:
Layout fixes

nirvn commented:
@nyalldawson , things behaving as they should under proj5, will test things out under proj6 whenever osgeo4w's qgis-dev builds include those commits. 

Fantastic efforts increasing reliability of QGIS under brand-new proj6. And thanks to @rouault for being so responsive throughout this fix sprint. I wouldn't want to run a gdal/proj environment pre all the changes and fixes done in the last 5-6 days.

On that note, would it be worth making soon-to-be-released PROJ 6.3 be the minimum version requirement?

nyalldawson committed:
Fix doxygen warning

nyalldawson committed:
Fix typo in CRS code

nyalldawson committed:
Prefer WKT2 when exporting CRS for use by GDAL

nyalldawson committed:
Don't cache WKT strings inside QgsCoordinateReferenceSystem

There's many different formats of WKT strings, so instead play it safe
and always export in the exact specified format instead of using a cached
value which may be a different format.

nyalldawson committed:
Update test results to use WKT2

nyalldawson committed:
Round trip test CRSes via WKT

nyalldawson committed:
Fix build, fix test on proj < 6

nyalldawson committed:
Add some proj6 test branches

nyalldawson commented:
Unrelated test failure
